File: app/src/main/java/ezike/tobenna/myweather/di/module/ApiModule.java
package ezike.tobenna.myweather.di.module;

import android.app.Application;

import com.squareup.moshi.Moshi;

import java.io.File;
import java.util.concurrent.TimeUnit;

import javax.inject.Named;
import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import ezike.tobenna.myweather.data.network.LiveDataCallAdapterFactory;
import ezike.tobenna.myweather.data.network.api.ApiService;
import ezike.tobenna.myweather.data.network.interceptors.ApiInterceptor;
import okhttp3.Cache;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.converter.moshi.MoshiConverterFactory;

/**
 * @author tobennaezike
 */
@Module(includes = ClientModule.class)
public class ApiModule {

    @Provides
    @Singleton
    static Cache provideCache(Application application) {
        long cacheSize = 10 * 1024 * 1024; // 10 MB
        File httpCacheDirectory = new File(application.getCacheDir(), "http-cache");
        return new Cache(httpCacheDirectory, cacheSize);
    }

    @Provides
    @Singleton
    static OkHttpClient provideOkHttpClient(Cache cache, @Named("connect") ApiInterceptor connectivityInterceptor,
                                            @Named("request") ApiInterceptor requestInterceptor) {

        HttpLoggingInterceptor logging = new HttpLoggingInterceptor();
        logging.setLevel(HttpLoggingInterceptor.Level.BODY);
        OkHttpClient.Builder httpClient = new OkHttpClient.Builder();
        httpClient.cache(cache);
        httpClient.addInterceptor(logging);
        httpClient.addNetworkInterceptor(requestInterceptor);
        httpClient.addInterceptor(connectivityInterceptor);
        httpClient.connectTimeout(30, TimeUnit.SECONDS);
        httpClient.readTimeout(30, TimeUnit.SECONDS);
        return httpClient.build();
    }

    @Provides
    @Singleton
    static Moshi provideMoshi() {
        return new Moshi.Builder().build();
    }

    @Singleton
    @Provides
    static ApiService provideApiService(Moshi moshi, OkHttpClient okHttpClient) {
        return new Retrofit.Builder()
                .baseUrl("http://api.apixu.com/v1/")
                .addConverterFactory(MoshiConverterFactory.create(moshi))
                .addCallAdapterFactory(new LiveDataCallAdapterFactory())
                .client(okHttpClient)
                .build()
                .create(ApiService.class);
    }
}

File: app/src/main/java/ezike/tobenna/myweather/di/module/ApiModule.kt
package ezike.tobenna.myweather.di.module

import android.app.Application
import com.squareup.moshi.Moshi
import com.squareup.moshi.kotlin.reflect.KotlinJsonAdapterFactory
import dagger.Module
import dagger.Provides
import ezike.tobenna.myweather.AppCoroutineDispatchers
import ezike.tobenna.myweather.BuildConfig
import ezike.tobenna.myweather.data.remote.api.ApiService
import ezike.tobenna.myweather.data.remote.interceptors.ConnectivityInterceptor
import ezike.tobenna.myweather.data.remote.interceptors.RequestInterceptor
import okhttp3.Cache
import okhttp3.OkHttpClient
import okhttp3.OkHttpClient.Builder
import okhttp3.logging.HttpLoggingInterceptor
import okhttp3.logging.HttpLoggingInterceptor.Level
import retrofit2.Retrofit
import retrofit2.converter.moshi.MoshiConverterFactory
import java.io.File
import javax.inject.Singleton

/**
 * @author tobennaezike
 */

@Module
object ApiModule {
    @Provides
    @JvmStatic
    @Singleton
    internal fun provideCache(application: Application): Cache {
        val cacheSize = 10 * 1024 * 1024.toLong() // 10 MB
        val httpCacheDirectory = File(application.cacheDir, "http-cache")
        return Cache(httpCacheDirectory, cacheSize)
    }

    @Provides
    @Singleton
    internal fun provideOkHttpClient(cache: Cache, connectivityInterceptor: ConnectivityInterceptor,
                                     requestInterceptor: RequestInterceptor,
                                     logger: HttpLoggingInterceptor): OkHttpClient {
        val httpClient = Builder()
        httpClient.cache(cache)
        httpClient.addInterceptor(logger)
        httpClient.addNetworkInterceptor(requestInterceptor)
        httpClient.addInterceptor(connectivityInterceptor)
        return httpClient.build()
    }

    @Provides
    @Singleton
    internal fun provideOkHttpLogger(): HttpLoggingInterceptor {
        return HttpLoggingInterceptor().apply {
            if (BuildConfig.DEBUG) {
                level = Level.BODY
            }
        }
    }

    @Provides
    @Singleton
    internal fun provideMoshi(): Moshi {
        return Moshi.Builder()
                .add(KotlinJsonAdapterFactory()).build()
    }

    @Singleton
    @Provides
    internal fun provideApiService(moshi: Moshi, okHttpClient: OkHttpClient): ApiService {
        return Retrofit
                .Builder()
                .baseUrl(BuildConfig.BaseUrl)
                .addConverterFactory(MoshiConverterFactory.create(moshi))
                .client(okHttpClient)
                .build()
                .create(ApiService::class.java)
    }

    @Provides
    fun provideDispatcher(): AppCoroutineDispatchers {
        return AppCoroutineDispatchers()
    }
}